File: knowledge-base-backend-master/kb-document/src/main/java/com/knowledge/base/document/service/DocumentService.java
package com.knowledge.base.document.service;

import com.baomidou.mybatisplus.core.metadata.IPage;
import com.baomidou.mybatisplus.extension.service.IService;
import com.knowledge.base.document.dto.AutoSaveDTO;
import com.knowledge.base.document.dto.DocumentDTO;
import com.knowledge.base.document.entity.Document;
import com.knowledge.base.document.vo.DocumentNeighborVO;
import com.knowledge.base.document.vo.DocumentVO;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

/**
 * Document Service interface
 *
 * <p>Designed following the Alibaba Java Coding Guidelines, provides document business logic operations</p>
 *
 * @author airwzz999
 * @since 1.0.0
 */
public interface DocumentService extends IService<Document> {

    /**
     * Creates a document
     *
     * @param documentDTO document information
     * @return document ID
     */
    Long createDocument(DocumentDTO documentDTO);

    /**
     * Updates a document
     *
     * @param documentDTO document information
     * @return whether successful
     */
    Boolean updateDocument(DocumentDTO documentDTO);

    /**
     * Updates the document summary (updates only the summary field, without full validation)
     *
     * @param documentId document ID
     * @param summary    summary content
     * @return whether successful
     */
    Boolean updateSummary(Long documentId, String summary);

    /**
     * Deletes a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean deleteDocument(Long documentId);

    /**
     * Queries a document by ID
     *
     * @param documentId document ID
     * @return document information
     */
    DocumentVO getDocumentById(Long documentId);

    /**
     * Views a document (increments the view count)
     *
     * @param documentId document ID
     * @return document information
     */
    DocumentVO viewDocument(Long documentId);

    /**
     * Paginated query of the document list
     *
     * @param current    current page
     * @param size       page size
     * @param categoryId category ID
     * @param keyword    search keyword
     * @param status     status
     * @param sortBy     sort field
     * @param sortOrder  sort direction
     * @return paginated document information
     */
    IPage<DocumentVO> pageDocuments(Long current, Long size, Long categoryId, Long teamId, String keyword, Integer status, String sortBy, String sortOrder, Long authorId);

    /**
     * Queries the previous and next document
     *
     * @param documentId current document ID
     * @return neighboring document information (prevId/prevTitle/nextId/nextTitle)
     */
    DocumentNeighborVO getDocumentNeighbors(Long documentId);

    /**
     * Uploads a document file
     *
     * @param file file
     * @return file path
     */
    String uploadDocumentFile(MultipartFile file);

    /**
     * Likes a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean likeDocument(Long documentId);

    /**
     * Unlikes a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean unlikeDocument(Long documentId);

    /**
     * Favorites a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean favoriteDocument(Long documentId);

    /**
     * Publishes a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean publishDocument(Long documentId);

    /**
     * Archives a document
     *
     * @param documentId document ID
     * @return whether successful
     */
    Boolean archiveDocument(Long documentId);

    /**
     * Cleans up stale knowledge graph nodes (document graph nodes in Neo4j left over after deletion in MySQL)
     *
     * @return number of nodes cleaned up
     */
    int cleanupGraphGhostNodes();

    /**
     * Uploads a file and parses it to create a document
     *
     * @param file uploaded file
     * @return Map containing documentId / title / fileUrl / fileSize / contentLength / contentPreview
     */
    Map<String, Object> uploadAndCreateDocument(MultipartFile file);

    /**
     * Rebuilds the knowledge graph for all published documents
     *
     * @return number of documents rebuilt
     */
    int rebuildAllGraphs();

    /**
     * Auto-saves a document (creates a new draft or updates an existing draft)
     *
     * <p>Key differences from createDocument/updateDocument:
     * <ul>
     *   <li>Title is optional; an empty title is auto-filled with "Untitled Document"</li>
     *   <li>Forces the status to draft (0), does not trigger RAG/KAG/ES indexing</li>
     *   <li>Only updates non-empty fields, avoiding overwriting existing data</li>
     * </ul></p>
     *
     * @param autoSaveDTO auto-save data
     * @return document ID (returns the new ID when creating, or the existing ID when updating)
     */
    Long autoSaveDocument(AutoSaveDTO autoSaveDTO);

    /**
     * Dismisses auto-saved drafts: marks all of the current user's drafts (status=0) as acknowledged,
     * so the restore prompt no longer appears afterward.
     */
    void dismissAutoSaveDrafts();
}

File: knowledge-base-frontend-master/src/services/document.service.ts
import { http } from './request';
import request from './request';
import { Document, DocumentFilter } from '@/types';

export const documentService = {
  // Get document list (paginated)
  // Backend returns IPage<DocumentVO> structure: { records, total, current, size }
  // Frontend expects DocumentListResponse structure: { list, total, page, pageSize }
  getDocuments: (filter: DocumentFilter) => {
    // Convert frontend parameter names to the names expected by the backend
    const params: Record<string, any> = {
      current: filter.page || 1,
      size: filter.pageSize || 10,
    };

    // Only add parameters supported by the backend
    if (filter.categoryId) params.categoryId = filter.categoryId;
    if (filter.teamId) params.teamId = filter.teamId;
    if (filter.keyword) params.keyword = filter.keyword;
    if (filter.status !== undefined) params.status = filter.status;
    if (filter.authorId) params.authorId = filter.authorId;

    // Add sort parameters
    if (filter.sortBy) params.sortBy = filter.sortBy;
    if (filter.sortOrder) params.sortOrder = filter.sortOrder;

    return http.get<any>('/document/documents/page', { params }).then((data) => ({
      list: data.records || [],
      total: data.total || 0,
      page: data.current || 1,
      pageSize: data.size || 10,
    }));
  },

  // Get document details
  getDocument: (id: string) => {
    return http.get<Document>(`/document/documents/${id}`);
  },

  // Get the previous and next document
  getDocumentNeighbors: (id: string) => {
    return http.get<{ prevId: string | null; prevTitle: string | null; nextId: string | null; nextTitle: string | null }>(`/document/documents/${id}/neighbors`);
  },

  // Create document
  // Backend returns Result<Long>, i.e. the document ID
  createDocument: (data: any) => {
    return http.post<{ id: string }>('document/documents', data);
  },

  // Update document (note: backend endpoint is PUT /documents, with id in the request body)
  updateDocument: (id: string, data: Partial<Document>) => {
    return http.put<Document>('/document/documents', { ...data, id });
  },

  // Auto-save document (create or update a draft, allows empty title, does not trigger indexing)
  autoSaveDocument: (data: {
    id?: number | string;
    title?: string;
    content?: string;
    summary?: string;
    categoryId?: number | string;
    teamId?: number | string;
    tags?: string;
  }) => {
    return http.post<{ id: string }>('/document/documents/autosave', data);
  },

  // Update only the document summary (uses a dedicated PATCH endpoint, skips full validation)
  updateSummary: (id: string, summary: string) => {
    return http.patch<boolean>(`/document/documents/${id}/summary`, { summary });
  },

  // Delete document
  deleteDocument: (id: string) => {
    return http.delete(`/document/documents/${id}`);
  },

  // Publish document (note: backend endpoint is PUT /documents/{id}/publish)
  publishDocument: (id: string) => {
    return http.put<Document>(`/document/documents/${id}/publish`);
  },

  // Archive document (note: backend endpoint is PUT /documents/{id}/archive)
  archiveDocument: (id: string) => {
    return http.put<Document>(`/document/documents/${id}/archive`);
  },

  // Like document
  likeDocument: (id: string) => {
    return http.post(`/document/documents/${id}/like`);
  },

  // Unlike document
  unlikeDocument: (id: string) => {
    return http.delete(`/document/documents/${id}/like`);
  },

  // Favorite document
  favoriteDocument: (id: string) => {
    return http.post(`/document/documents/${id}/favorite`);
  },

  // View document (increments view count)
  viewDocument: (id: string) => {
    return http.get<Document>(`/document/documents/${id}/view`);
  },

  // Note: category-related methods have been moved to categoryService, use categoryService instead

  // Upload document file
  uploadDocumentFile: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    // Use the postForm method to handle file upload
    return (http as any).postForm<string>('/document/documents/upload', formData);
  },

  // Upload a file and parse it to create a document
  uploadAndParseDocument: (file: File): Promise<{
    documentId: string;
    title: string;
    fileUrl: string;
    fileSize: number;
    contentLength: number;
    contentPreview: string;
  }> => {
    const formData = new FormData();
    formData.append('file', file);
    return (http as any).postForm('/document/documents/upload/parse', formData);
  },

  // Export document to PDF (get download link)
  exportDocumentToPdf: (documentId: string) => {
    return http.get<string>(`/document/documents/${documentId}/export-pdf`);
  },

  // Download document PDF (direct download)
  downloadDocumentPdf: (documentId: string) => {
    return (http as any).download(`/document/documents/${documentId}/download-pdf`);
  },

  // Batch export documents
  batchExportDocuments: async (documentIds: string[], format: 'pdf' | 'markdown') => {
    const response: any = await request.post('/document/documents/batch-export', { documentIds, format }, {
      responseType: 'blob',
      _download: true,
    } as any);

    // Check whether the response is a JSON error (rather than a ZIP file)
    const contentType = response.headers?.['content-type'] || '';
    if (contentType.includes('application/json')) {
      // Read the error message
      const text = await response.data.text();
      const errorData = JSON.parse(text);
      throw new Error(errorData.message || 'Export failed');
    }

    const contentDisposition = response.headers?.['content-disposition'] || '';
    let filename = `documents_export_${new Date().toISOString().slice(0, 10)}.zip`;
    const match = contentDisposition.match(/filename\*=UTF-8''([^"';]+)/);
    if (match) {
      try { filename = decodeURIComponent(match[1]); } catch { filename = match[1]; }
    }
    const blob = new Blob([response.data], { type: 'application/zip' });
    const link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(link.href);
  },

  // Create share link
  createShare: (data: {
    documentId: string;
    shareType?: number;
    expireType?: number;
    expireTime?: string;
    accessLimit?: number;
    requirePassword?: number;
    password?: string;
    description?: string;
  }) => {
    return http.post<ShareVO>('/document/documents/share', data);
  },

  // Get share info
  getShareInfo: (shareId: string) => {
    return http.get<ShareVO>(`/document/documents/share/${shareId}`);
  },

  // Access share link
  accessShare: (shareId: string, password?: string) => {
    const params = password ? { password } : {};
    return http.post<number>(`/document/documents/share/${shareId}/access`, null, { params });
  },

  // Get all shares of a document
  getDocumentShares: (documentId: string) => {
    return http.get<ShareVO[]>(`/document/documents/${documentId}/shares`);
  },

  // Get my share list
  getMyShares: () => {
    return http.get<ShareVO[]>('/document/documents/share/my');
  },

  // Delete share link
  deleteShare: (shareId: string) => {
    return http.delete(`/document/documents/share/${shareId}`);
  },

  // Update share settings
  updateShare: (shareId: string, data: any) => {
    return http.put(`/document/documents/share/${shareId}`, data);
  },

  // ========== Public share endpoints (no login required, skipAuth) ==========

  // Publicly get share info
  getPublicShareInfo: (shareId: string) => {
    return http.get<ShareVO>(`/document/share/${shareId}`, { skipAuth: true } as any);
  },

  // Publicly verify share password
  verifyPublicShare: (shareId: string, password?: string) => {
    const params: Record<string, string> = {};
    if (password) params.password = password;
    return http.post<boolean>(`/document/share/${shareId}/verify`, null, {
      params,
      skipAuth: true,
    } as any);
  },

  // Publicly access share (get document content)
  accessPublicShare: (shareId: string, password?: string) => {
    const params: Record<string, string> = {};
    if (password) params.password = password;
    return http.post<any>(`/document/share/${shareId}/access`, null, {
      params,
      skipAuth: true,
    } as any);
  },

  // ========== Auto-save history ==========

  // Get the auto-save history snapshot list of a document
  getAutoSaveHistory: (documentId: string, page = 1, pageSize = 20) => {
    return http.get<any>(`/document/documents/${documentId}/autosave-history`, {
      params: { current: page, size: pageSize },
    }).then((data: any) => ({
      list: data.records || [],
      total: data.total || 0,
      page: data.current || 1,
      pageSize: data.size || 20,
    }));
  },

  // Get a single auto-save snapshot's details (including full content)
  getAutoSaveSnapshot: (documentId: string, snapshotId: string) => {
    return http.get<any>(`/document/documents/${documentId}/autosave-history/${snapshotId}`);
  },

  // Dismiss auto-saved drafts (marks all of the current user's drafts as acknowledged, so the restore prompt no longer appears)
  dismissAutoSaveDrafts: () => {
    return http.put<boolean>('/document/documents/autosave/dismiss');
  },
};

export interface ShareVO {
  shareId: string;
  shareUrl: string;
  documentId: string | number;
  title: string;
  shareType: number;
  shareTypeDesc: string;
  expireType: number;
  expireTime: string;
  expired: boolean;
  requirePassword: boolean;
  sharerName: string;
  shareTime: string;
  accessCount: number;
  description: string;
}

export default documentService;
